### PR TITLE
Refactor tracing + add code for state analysis

### PR DESCRIPTION
**Pull request by JayFoxRox**
Based on #2 

Run like `XBOX_IF=none ./decode-state.py out/command4647 16`

### CODE REVIEW

File: Trace.py [hunk 1]
```diff
@@ -202,6 +202,18 @@ def DumpSurfaces(self, xbox, data, *args):
...
 
 
+    # We always dump twice from what we expect
+    if True:
+      def out(suffix, contents):
+        with open(("out/command%d_" % self.commandCount)+ suffix, "wb") as f:
+          f.write(contents)
+      buffer_size = pitch * height * 2
+      out("mem.bin", xbox.read(0x80000000 | offset, buffer_size))
+      out("pgraph.bin", dumpPGRAPH(xbox))
```
Comment (JayFoxRox) [inline Trace.py]: Also need to dump PFB

File: nv_tiles.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+  x >>= 8
+  y >>= bankshift - 8
+  iaddr = 0
+  baddr = y * (pitch >> 8) + x
+  part = 0
+  tag = 0
+
+  fb_type = pfb_type(chipset)
+  if fb_type == PFB_NV10:
```
Comment (JayFoxRox) [inline nv_tiles.py]: This codepath can probably be removed. If this is bad, blame it all on mooch

File: auto-swizzle.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+#FIXME: Load PGRAPH state
+#FIXME: Load pixel data
+
+#FIXME: decode texture swizzled / unswizzled, rounded dimensions / clean dimensions
+
+path = sys.argv[1]
+
+mem = load_out(path, "mem.bin")
+pgraph = load_out(path, "pgraph.bin")
```
Comment (JayFoxRox) [inline auto-swizzle.py]: Load PFB

**Comment by JayFoxRox** [comment]
I had a crash while running 0x1D70 in the XDK-launcher (with PFB dumping enabled).

Might be a bug in #2. This could be bad and should be looked into.

File: nv_tiles.py (hunk 2 above)
Comment (MoochMcGee) [reply nv_tiles.py]: lol fair enough

File: nv_tiles.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    self.partbits = 0
+    self.colbits_lo = 0
+    self.burstbits = 0
+
+    if True:
+      #FIXME: Add https://github.com/envytools/envytools/blob/master/hwtest/nv10_tile.cc#L35
+      self.mcbits = 2
+      self.partbits = 2
+      self.colbits_lo = 2
+      self.burstbits = 0
```
Comment (JayFoxRox) [inline nv_tiles.py]: This is still incomplete!

File: nv_tiles.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+
+def tile_translate_addr(chipset, pitch, address, mode, bankoff, mcc):
```
Comment (JayFoxRox) [inline nv_tiles.py]: mode is always 1 pre PFB_NV40 (?) according to mwk

**Comment by JayFoxRox** [comment]
Bloodwake uses `Surface type non-swizzled; anti-aliasing: center-corner-2` - that should help a lot!

**Comment by JayFoxRox** [comment]
Fog info should be dumped. It seems to be a problem with Tony Hawks American Wasteland (and possibly other games - maybe even Halo).

File: decode-state.py [hunk 6]
```diff
@@ -35,6 +36,279 @@ def align_up(v, alignment):
...
+    gcc2 = "vec3(dot(%s, %s))" % (a, b)
+    gcc3 = "%s * %s" % (c, d)
+    gcc4 = "vec3(dot(%s, %s))" % (c, d)
+    gcc5 = "%s + %s" % (gcc1, gcc3)
+    gcc6 = "mix(%s, %s, SPARE0_NV.a & %s)" % (gcc3, gcc1, mux_bit_str)
+  else:
+    gcc1 = "%s * %s" % (a, b)
+    gcc2 = "%s * %s" % (c, d)
+    gcc3 = "%s + %s" % (gcc1, gcc2)
+    gcc4 = "mix(%s, %s, SPARE0_NV.a & %s)" % (gcc2, gcc1, mux_bit_str)
```
Comment (JayFoxRox) [inline decode-state.py]: don't resolve these vars here.. just put the logic of the used ones into the code

**Comment by JayFoxRox** [comment]
When `run_fifo` has run, PGRAPH might still be busy? we should wait for idle probably.

**Comment by JayFoxRox** [comment]
Plans for this:

- Do callback refactor
- Disable tiles in tracing
- Then fix xboxpy API
- Split tool into seperate PRs, just taking one function at a time

File: README.md [hunk 7]
```diff
@@ -53,6 +53,6 @@ If this tool does not work, please retry a couple of times.
 
 ---
 
-(c)2018 XboxDev maintainers
+**&copy; 2018 XboxDev maintainers**
```
Comment (JayFoxRox) [inline README.md]: Use (C)

**Comment by JayFoxRox** [comment]
RDI reads are slow and there should be an optimized version to do that access

**Comment by JayFoxRox** [comment]
Split into #9 and #4

**Comment by JayFoxRox** [comment]
Keeping repo undeleted until everything has been merged (this repo is already outdated at time of writing).

**Pull request closed without merging**